feat(tests): make it easier to bootstrap PHPUnit

hypeJunction opened the pull request:
Deprecates /engine/tests/phpunit/bootstrap.php in favour of composer autoloader.
Adds \Elgg\TestCase class that should be extended by individual test cases.

hypeJunction committed:
feat(tests): make it easier to bootstrap PHPUnit

Deprecates /engine/tests/phpunit/bootstrap.php in favour of composer autoloader.
Adds \Elgg\TestCase class that should be extended by individual test cases.

Adds global constant to phpunit.xml which can be used to sniff if
application is bootstrapped for testing.

Allowed ajax view register is now maintained within the ajax service
instead of $_ELGG global.

view_path configuration is no longer stored in $_ELGG global. The value is
now stored as a variable in views service.

hypeJunction commented:
The motivation is to eventually allow plugins to use composer to require elgg as a dev dependency and bootstrap themselves using `vendor/autoload.php` and run tests extending `\Elgg\TestCase`. Currently this is messy trying to include core bootstrap that expects a certain directory structure.

hypeJunction commented:
@mrclay I think this should also help with caching of globals. We can subsequently bootstrap most commonly used services, e.g. hooks, session etc in the constructor to avoid all the duplication. Individual tests can overwrite values if they need to.

mrclay commented:
> should also help with caching of globals

Probably not. `$app->loadCore()` on repeated calls bails out, so the fresh `$sp` we make never gets injected into `_elgg_services()`. More coming...

mrclay commented on engine/tests/phpunit/Elgg/TestCase.php:
```diff
@@ -0,0 +1,165 @@
+		 */
+		global $CONFIG;
+		$CONFIG = (object) [
+					'Config_file' => false,
```
1 Indent

mrclay commented on engine/tests/phpunit/Elgg/TestCase.php:
```diff
@@ -0,0 +1,165 @@
+		$app = new Application($sp);
+		Application::$_instance = $app;
+		$app->loadCore();
+
```
Here add:
```php
// loadCore bails on repeated calls, so we need to manually inject this to make
// sure it happens before each test.
_elgg_services($sp);
```

mrclay replied:
With this we *should* have fewer leftover state problems, not surely not a complete fix.

mrclay commented on engine/tests/phpunit/Elgg/TestCase.php:
```diff
@@ -0,0 +1,165 @@
+		 * testable without global state.
+		 */
+		global $CONFIG;
+		$CONFIG = (object) [
```
Can we require use a common script to keep these values in sync with the legacy bootstrap or would that defeat the purpose?

mrclay replied:
Maybe most of this logic should go in a static method and we can use that from the old bootstrap?

hypeJunction replied:
We can require the old boostrap. Once the class is autoloaded the bootstrap path should be predictable.

mrclay commented:
I'm overall fine with the direction, but don't know testing culture well enough to know if it's wise. We're basically forcing this particular bootstrap strategy before every test. [shrugs]

hypeJunction commented:
I think this increased the test speed by a few seconds. But I am not sure either. Maybe @Srokap has some input?

hypeJunction commented:
@mrclay See how you like this. Our API is in a good shape and no longer relies on global `$CONFIG`, so I removed it from tests. I refactored uses of `view_path` and `allowed_ajax_views` so we no longer need `$_ELGG` (we should get rid of it altogether soon).

What else we can do:
- Get rid of `\Elgg\TestCase::config()` and refactor tests to use `_elgg_services()`
- Add `__set()` and `__get()` to `\Elgg\TestCase` and proxy that to `_elgg_services()` so our test case values will be in sync with SP

hypeJunction commented:
This is now fully BC. Added a test case to test that test case is BC.

mrclay commented on composer.json:
```diff
@@ -62,7 +62,10 @@
-        }
+        },
+		"files": [
+			"engine/tests/phpunit/bootstrap.php"
```
Wouldn't this run this file anytime the autoloader is used?

mrclay replied:
I see you sniff for a constant and bail if not testing. Still, autoloading this file feels dirty.

hypeJunction replied:
We could require it from TestCase, but that doesn't seem clean either

mrclay replied:
How about we leave the duplicate code in TestCase and add cross-referencing comments to keep both in sync? We'll rarely have to touch this and the duplication will be gone in master.

hypeJunction replied:
We still need to require bootstrap somehow/somewhere to not break tests pending in PRs or already in master.

hypeJunction replied:
I don't think duplicating logic in TestCase helps us. If the suite doesn't instantiate TestCase composer won't include the class file.

mrclay replied:
> to not break tests pending in PRs or already in master

Short term annoyance. I can take care of a lot of this.

hypeJunction replied:
To clarify, you are saying remove composer autoload, require from TestCase? Should I keep the constant in phpunit?

mrclay replied:
Yes and yes.

mrclay commented:
👍  for removing `$_ELGG->view_path`. `$_ELGG` was a temporary hack to to make refactoring safer for 2.0.

mrclay commented on phpunit.xml:
```diff
@@ -16,4 +16,7 @@
 		</blacklist>
 	</filter>
+	<php>
+		<const name="PHPUNIT_ELGG_TESTING_APPLICATION" value="true"/>
```
...I do like this though.

hypeJunction commented:
@mrclay Updated. Renamed some of the functions to better reflect their purpose.

mrclay commented on engine/classes/Elgg/Ajax/Service.php:
```diff
@@ -261,4 +267,32 @@ public function appendDeps($hook, $type, $response, $params) {
+	 * Returns an array of views allowed for ajax calls
+	 * @return string[]
+	 */
+	public function getViews() {
```
Plan for later: convert this to isRegisteredView($view) since that's all we need it for.

hypeJunction replied:
IMO, we may end up needing a list of views as we did with other services

mrclay commented on engine/tests/phpunit/Elgg/TestCase.php:
```diff
@@ -0,0 +1,151 @@
+	 * Returns default testing configuration array
+	 * @return array
+	 */
+	public static function getTestingConfigArray() {
```
👍

mrclay commented on engine/tests/phpunit/Elgg/TestCaseTest.php:
```diff
@@ -0,0 +1,18 @@
+	 * Test that legacy bootstrap has been autoloaded and
+	 * stay BC with older test cases
+	 */
+	public function testIsBoostrapped() {
```
Cool

hypeJunction replied:
Dang. I misspell bootstrap every time I type it.

mrclay replied:
I didn't notice!

mrclay commented on engine/tests/phpunit/bootstrap.php:
```diff
@@ -1,72 +1,45 @@
  * @return \Elgg\Application
+ * @deprecated 2.3 Use elgg() instead
  */
 function _elgg_testing_application(\Elgg\Application $app = null) {
```
We're keeping these private functions around in case of what?

hypeJunction replied:
We can toss them. I thought if we are keeping the bootstrap, we might as well keep the functions for the time being.

mrclay commented:
LGTM

hypeJunction commented:
I think we should go ahead and merge. I don't think anyone else on the team is bothered enough.
